mobius-core/src/main/java/com/spotify/mobius/FilteringEventSource.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * -\-\-
+ * Mobius
+ * --
+ * Copyright (c) 2017-2020 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+package com.spotify.mobius;
+
+import javax.annotation.Nonnull;
+
+/**
+ * An {@link EventSource} that also provides the functionality that the values it emits may be used
+ * as a filter for another Event Source.
+ *
+ * @param <T> The event class
+ */
+public interface FilteringEventSource<T> extends EventSource<T> {
```

Comment:
I'm not a huge fan of calling this "filtering". I associate that word much stronger with just ignoring some events, and not starting/stopping something based on a condition. In Rx (JS/Java/C#/etc) this kind of behavior is referred to as "switching" (technically it's a special case of switchMap). I don't know if it's the perfect name for us, but I believe "filter" is usually one of the most well defined operators when it comes to reactive functional programming, and the classes here that use the name "filter" aren't following that definition.

Reply:
Hm, true, this doesn't filter on a per-event basis. Happy to rename it to something like `SwitchMappingEventSource` ? or better name if you have any suggestions?

Reply:
I don't think this interface is needed. What seems to be happening is:

1. EventSource A produces data that shouldn't always be passed along.
2. EventSource B produces some other data that determines whether A's data should be thrown away.

So what's needed is like an operator like
```
   (EventSource<A>, EventSource<B>, Function<B, Boolean>) -> EventSource<A>
```
I think it's like a 'gate' in electronics, right? Only open if the signal on 'B' evaluates to `true`. Maybe that's what it should be called? Something like:

```
   EventSource<A> gated = EventSources.gatedBy(eventSourceB, conditionOnB).apply(eventSourceA);
```